dev/src/bulk-writer.ts

```diff
@@ -89,134 +159,77 @@ enum BatchState {
  * @private
  */
 class BulkCommitBatch extends WriteBatch {
-  /**
-   * The state of the batch.
-   */
-  private state = BatchState.OPEN;
-
   // The set of document reference paths present in the WriteBatch.
   readonly docPaths = new Set<string>();
 
   // An array of pending write operations. Only contains writes that have not
   // been resolved.
-  private pendingOps: Array<Deferred<BatchWriteResult>> = [];
-
-  constructor(firestore: Firestore, readonly maxBatchSize: number) {
-    super(firestore);
-  }
+  private pendingOps: Array<BulkWriterOperation> = [];
 
-  has(documentRef: firestore.DocumentReference): boolean {
+  has(documentRef: firestore.DocumentReference<unknown>): boolean {
     return this.docPaths.has(documentRef.path);
   }
 
-  markReadyToSend(): void {
-    if (this.state === BatchState.OPEN) {
-      this.state = BatchState.READY_TO_SEND;
-    }
-  }
-
-  isOpen(): boolean {
-    return this.state === BatchState.OPEN;
-  }
-
-  isReadyToSend(): boolean {
-    return this.state === BatchState.READY_TO_SEND;
-  }
-
-  async bulkCommit(): Promise<void> {
-    assert(
-      this.state === BatchState.READY_TO_SEND,
-      'The batch should be marked as READY_TO_SEND before committing'
-    );
-    this.state = BatchState.SENT;
+  async bulkCommit(options: {requestTag?: string} = {}): Promise<void> {
+    const tag = options?.requestTag ?? requestTag();
 
     // Capture the error stack to preserve stack tracing across async calls.
     const stack = Error().stack!;
 
-    let results: BatchWriteResult[] = [];
+    let response: api.IBatchWriteResponse;
     try {
+      logger(
+        'BulkCommitBatch.bulkCommit',
+        tag,
+        `Sending next batch with ${this._opCount} writes`
+      );
       const retryCodes = getRetryCodes('batchWrite');
-      const response = await this._commit<
+      response = await this._commit<
         api.BatchWriteRequest,
         api.BatchWriteResponse
-      >({retryCodes, methodName: 'batchWrite'});
+      >({retryCodes, methodName: 'batchWrite', requestTag: tag});
+    } catch (err) {
+      // Map the failure to each individual write's result.
+      const ops = Array.from({length: this.pendingOps.length});
+      response = {
+        writeResults: ops.map(() => {
+          return {};
+        }),
+        status: ops.map(() => err),
+      };
+    }
 
-      results = response.writeResults.map((result, i) => {
-        const status = response.status[i];
+    for (let i = 0; i < (response.writeResults || []).length; ++i) {
+      // Since delete operations currently do not have write times, use a
+      // sentinel Timestamp value.
+      // TODO(b/158502664): Use actual delete timestamp.
+      const DELETE_TIMESTAMP_SENTINEL = Timestamp.fromMillis(0);
+
+      const status = (response.status || [])[i];
```

Comment:
If the code is `undefined`, we pass it as `undefined` to the customer. That seems to be the reasonable thing to do, since we don't want to make up our own status code. But as you said, the status code should always be set.